docs/dev/writing-code.md

```diff
@@ -192,18 +201,195 @@ TODO: document when to use future vs callback
+        PyObject *capsule = PyCapsule_New(binding, on_capsule_destroyed_fn);
+        return capsule;
+    }
+    ```
+*   Things stay alive because:
+    *   The user's `elg` variable keeps the `EventLoopGroup` object alive.
+    *   Member variable `EventLoopGroup._binding` keeps the `PyCapsule` alive.
+    *   The `PyCapsule` keeps the `struct event_loop_group_binding` alive.
+    *   The `event_loop_group_binding.native` pointer is a "strong reference" that keeps
+        `struct aws_event_loop_group` alive.
+
+
+Destruction goes like this (it's actually more complex, we'll cover that later):
+*   When the user's Python code has no references to `elg`, the `EventLoopGroup` instance...
+*   The garbage collector cleans up the `EventLoopGroup`,
+    and the `PyCapsule` referenced by `EventLoopGroup._binding`.
+*   The `PyCapsule`'s destructor function runs, which looks something like:
+    ```C
+    static on_capsule_destroyed_fn(PyObject *capsule) {
+        struct event_loop_group_binding *binding = PyCapsule_GetPointer(capsule);
+
+        // release reference to underlying implementation
+        aws_event_loop_group_release(binding->native)
+
+        // free binding struct's memory
+        aws_mem_release(binding);
+    }
+    ```
+*   IF nothing else has a strong reference to `struct aws_event_loop_group`:
+    *   then it begins its shutdown process, and its memory is cleaned up when shutdown completes.
+*   ELSE something else has a strong reference to `struct aws_event_loop_group`:
+    *   so it won't begin its shutdown until the last reference is released.
+
+Note: In the past, the aws-c libs didn't have any reference counting for its C structs.
+You will still find older code in our Python bindings that tries to keep the entire dependency trees
+of Python objects alive via `Py_INCREF(x)` (TODO: remove needless complexity).
+You can't always look at existing code to see "the right way" of doing things.
+
+## A More Complex Example
+
+The sample above is simplified, it only shows Python calling into C.
+But `aws_event_loop_group` has a callback that fires when it finishes shutting down.
+That means C needs to call into Python AFTER the Python `EventLoopGroup` object has been cleaned up.
+
+For C to call into Python, it must reference a Python object
+(the function itself, or an object with a member function).
+This means our binding needs to store a strong reference and
+keep that Python object alive until the callback has fired.
+
+### The Wrong Way to Build it
+
+You MUST NOT create a reference cycle!
+
+You might be tempted to give `event_loop_group_binding` a strong reference to the `EventLoopGroup` instance.
+Then C could simply call a private member function like `EventLoopGroup._on_shutdown_complete()`.
+But this design creates a reference cycle (see image below):
+
+![Diagram of bad binding with reference cycle](./binding-bad-cycle.svg)
+
+### Option 1 - Pass Callbacks to C
+
+Most of our bindings work like this:
```

Comment:
Question: Should this diagram have an options/builder that points to both the EventLoopGroup and the Callable? From my current understanding, generally a builder/settings class or similar is how we set the callable and create the EventLoopGroup.

Reply:
Yeah, there's a ton more complexity that goes into hooking up real callbacks.

I'm just trying to illustrate that we need to avoid circular references, and keep out any details I can which would distract from that.